Add AppInstallPage

danielfmiranda opened the pull request:
# Description

<!-- Describe the PR here -->

Link to sample test Youtube page: https://foundation-s-tiktok-pag-1mbxqc.mofostaging.net/en/youtuberegrets/
Link to sample test TikTok page: https://foundation-s-tiktok-pag-1mbxqc.mofostaging.net/en/tiktok/
Link to Figma File: [Figma](https://www.figma.com/file/zMGcZQtSI2kqp4pgRqrOlu/TikTok-Reporter-(app-landing-page-template)?type=design&node-id=17-13998&mode=design&t=aXPGuBeOtyj1WeLj-0)
Related PR: https://github.com/MozillaFoundation/foundation.mozilla.org/issues/12036

This PR introduces a new page model `AppInstallPage` which is planned to be used on the upcoming [TikTok Reporter App page](https://www.figma.com/file/zMGcZQtSI2kqp4pgRqrOlu/TikTok-Reporter-(app-landing-page-template)?type=design&node-id=64-13208&mode=design&t=cXynojWLpUI1UNNM-0), and to also recreate the [Youtube Regrets Reporter Extension page](https://foundation.mozilla.org/en/youtube/regretsreporter/).

The ask for this page was to have the hero section of the regrets reporter page (with some minor design changes), and the "body/cta" section of a campaign page.

To achieve this I made the `AppInstallPage` inherit from our existing `CampaignPage` model, so they can share the same `body`/`cta` fields. 

I also had `app_install_page.html` extend `campaign_page.html`, with some minor updates to accommodate requests from design.

┆Issue is synchronized with this [Jira Story](https://mozilla-hub.atlassian.net/browse/TP1-345)


# Screenshots

## Figma Mockup:

![image](https://github.com/MozillaFoundation/foundation.mozilla.org/assets/18314510/40370716-3c96-4ea6-923d-97f6e8c4aefc)

## Review App:

**Please note that:**
- I uploaded button icons that do not match the figma mock up,  this is not an error.
- The circular "share" button underneath the download buttons in the figma mockup is missing on the review app, this is going to be added at a later date. 

![Screenshot 2024-03-29 at 00-58-10 TikTok](https://github.com/MozillaFoundation/foundation.mozilla.org/assets/18314510/72d00b1f-cbd6-457a-8bae-96c9c0b4794f)

danielfmiranda commented on network-api/networkapi/wagtailpages/templates/wagtailpages/campaign_page.html:
```diff
@@ -58,16 +58,18 @@ <h1 class="tw-h1-heading {% if uses_menu %}tw-hidden medium:tw-block{% endif %}"
                     {% block cta %}
                         {% if page.cta != None %}
-                            <div class="cta tw-pb-4
+                            <div class="cta tw-pb-4 tw-mx-auto
```
This `tw-mx-auto` class keeps the CTA form centered on mobile and tablet breakpoints

danielfmiranda commented on network-api/networkapi/wagtailpages/templates/wagtailpages/campaign_page.html:
```diff
@@ -58,16 +58,18 @@ <h1 class="tw-h1-heading {% if uses_menu %}tw-hidden medium:tw-block{% endif %}"
-                                <div class="narrow-sticky-button-container hidden">
-                                    <a href="#cta-anchor" class="tw-btn-primary tw-block">{% trans "TAKE ACTION" context "Sticky button in mobile view on campaign pages" %}</a>
-                                </div>
+                                {% block cta_sticky_button %}
```
Design requested that these pages do not have the sticky "take action" button that renders at the bottom of the page on mobile/tablet screens.

Adding this new `{% block cta_sticky_button %}` allows us to override this block on `app_install_page.html`

kristinashu commented:
So amazing, thank you @danielfmiranda! Everything is looking great, just a one question and comment:

1) I have a question about the CTA since it's hard to tell what it is in the review app - is the CTA snippet a petition or a newsletter signup? We need to have a newsletter sign up but for future proofing, it might be best if it could be either newsletter signor or petition (like on the existing banner pages).

2) I'm not sure how the top section is being resized but it seems too tall at some browser sizes. It should have a max height and resize like the top section of the Articles pages. For example https://foundation.mozilla.org/en/insights/trustworthy-ai-whitepaper/ 

Here is a screen cap comparing the two:
<img width="2555" alt="image" src="https://github.com/MozillaFoundation/foundation.mozilla.org/assets/1068783/43cbc805-7358-4c5d-aea8-91dbc9784b59">

I worry that if we let the image get too big than the resolution won't be good and the graphic won't crop as well. Example with the almost final graphic https://foundation-s-tiktok-pag-qneczb.mofostaging.net/en/tiktok-reporter/

mmmavis commented on network-api/networkapi/wagtailpages/templates/wagtailpages/fragments/app_install_hero.html:
```diff
@@ -0,0 +1,49 @@
+
+{% image page.hero_background original as hero_background %}
+
+<div class="app-install-page-hero tw-py-0 tw-flex tw-flex-col tw-justify-center tw-h-auto tw-min-h-[25em] tw-bg-center tw-bg-cover medium:tw-py-24 medium:tw-h-[65vh] large:tw-h-[75vh]" style="background-image:url('{{ hero_background.url }}')">
```
Instead of using inline `style` for background image, let's use Tailwind class instead.

mmmavis reviewed (commented):
@danielfmiranda Thanks for the PR. I noticed that the two buttons on the hero need some adjustments (general sizing & alignment between the icon and button text). Button hover state is grey which is something new. If this is part of the new addition to the styleguide, we should create a Tailwind utility class for it.

mmmavis reviewed (changes requested):
Please see my previous comments.

robdivincenzo commented on network-api/networkapi/wagtailpages/templates/wagtailpages/fragments/responsive_screen_wide_image.html:
```diff
@@ -0,0 +1,60 @@
+      Assuming low probability for a high res screen (since it's an outdated browser).
+    {% endcomment %}
+        src="{{ img_1200.url }}"
+        alt=""
```
Might be good to include the alt text here in case the hero isn't always decorative.

danielfmiranda replied:
Great idea @robdivincenzo! This has been updated in the latest commit, please see [this comment](https://github.com/MozillaFoundation/foundation.mozilla.org/pull/12122/files#r1550306674) for more details.

robdivincenzo commented on network-api/networkapi/wagtailpages/templates/wagtailpages/fragments/app_install_hero.html:
```diff
@@ -0,0 +1,63 @@
+{% load i18n wagtailcore_tags wagtailimages_tags %}
+
+<!-- {% image page.hero_background original as hero_background %} -->
```
Could remove stray comment while in here.

danielfmiranda replied:
Great catch @robdivincenzo! This has been removed

robdivincenzo commented on network-api/networkapi/wagtailpages/pagemodels/app_install.py:
```diff
@@ -0,0 +1,110 @@
+    button1_text = models.CharField(
+        max_length=50,
+        help_text="Text for Button 1",
+        null=True,
+        blank=True,
+    )
+    button1_download_link = models.URLField(
+        max_length=255,
+        help_text="Download link for Button 1",
+        null=True,
+        blank=True,
+    )
+    button1_image = models.ForeignKey(
+        "wagtailimages.Image",
+        null=True,
+        blank=True,
+        on_delete=models.SET_NULL,
+        related_name="+",
+        help_text="Image for Button 1",
+    )
+    button2_text = models.CharField(
+        max_length=50,
+        help_text="Text for Button 2",
+        null=True,
+        blank=True,
+    )
+    button2_image = models.ForeignKey(
+        "wagtailimages.Image",
+        null=True,
+        blank=True,
+        on_delete=models.SET_NULL,
+        related_name="+",
+        help_text="Image for Button 2",
+    )
+    button2_download_link = models.URLField(
+        max_length=255,
+        help_text="Download link for Button 2",
+        null=True,
+        blank=True,
+    )
```
Seems to me like these buttons can be condensed down to a re-usable StreamField block. Do we have one? That approach would also help reduce duplicated code in `network-api/networkapi/wagtailpages/templates/wagtailpages/fragments/app_install_hero.html` by allowing us to loop through the data.

danielfmiranda replied:
Thanks @robdivincenzo! This is an excellent idea, I have created the block [AppInstallDownloadButtonBlock](https://github.com/MozillaFoundation/foundation.mozilla.org/blob/bb8fe7a0290ccf70105cbe4d384c1786b185c800/network-api/networkapi/wagtailpages/pagemodels/customblocks/app_install_download_button_block.py) and am now using it in the latest commit. 👍

robdivincenzo commented on network-api/networkapi/wagtailpages/templates/wagtailpages/fragments/app_install_hero.html:
```diff
@@ -0,0 +1,63 @@
+                        {% with download_btn_classes="tw-max-w-max tw-btn tw-bg-white hover:tw-bg-gray-20 tw-transition-colors tw-border-black tw-border-2" btn_image_classes="tw-mr-2 tw-inline" %}
+                            <div class="download-btn-wrapper">
+                                <div class="btn-container tw-mt-2 tw-flex tw-flex-wrap tw-gap-8">
+                                    {% if page.button1_text and page.button1_download_link %}
+                                        <a class="{{ download_btn_classes }}" href="{{ page.button1_download_link }}" target="_blank">
+                                            {% if page.button1_image %}
+                                                {% image page.button1_image original as btn1_image %}
+                                                <img src="{{ btn1_image.url }}" alt="{{ page.button1_image_alt }}" class="{{ btn_image_classes }}" width="24" height="24">
+                                            {% endif %}
+                                            {{ page.button1_text }}
+                                        </a>
+                                    {% endif %}
+                                    {% if page.button2_text and page.button2_download_link %}
+                                        <a class="{{ download_btn_classes }}" href="{{ page.button2_download_link }}" target="_blank">
+                                            {% if page.button2_image %}
+                                                {% image page.button2_image original as btn2_image %}
+                                                <img src="{{ btn2_image.url }}" alt="{{ page.button2_image_alt }}" class="{{ btn_image_classes }}" width="24" height="24">
+                                            {% endif %}
+                                            {{ page.button2_text }}
+                                        </a>
+                                    {% endif %}
```
If we re-work the button data, this is where we can condense the code down to a loop and repeat less

danielfmiranda replied:
This has been updated in the latest commit to use the new `AppInstallDownloadButtonBlock` and to loop through them. Thanks for the suggestion Rob!

robdivincenzo reviewed (changes requested):
@danielfmiranda really great work on this! I have one small comment on the approach, wondering if we can condense/optimize the buttons with a re-usable common block. I'd like to hear your thoughts on that. And then also some very minor comments (alt text / stray comment)

danielfmiranda commented on network-api/networkapi/templates/fragments/responsive_screen_wide_image.html:
```diff
@@ -55,6 +55,6 @@
     {% endcomment %}
         src="{{ img_1200.url }}"
-        alt=""
+        alt="{% trans "Hero Image" %}"
```
Hi @robdivincenzo! 

This template is currently only being used as a hero image for the following templates:
- `custom_hero.html`
- `publication_hero.html`
- `app_install_hero.html`

Based on your feedback [here](https://github.com/MozillaFoundation/foundation.mozilla.org/pull/12122#discussion_r1548161649), I have added the generic alt text of "hero image" so this `alt` field is not empty 👍

danielfmiranda commented on network-api/networkapi/wagtailpages/templates/wagtailpages/fragments/app_install_hero.html:
```diff
@@ -0,0 +1,49 @@
+
+{% image page.hero_background original as hero_background %}
+
+<div class="app-install-page-hero tw-py-0 tw-flex tw-flex-col tw-justify-center tw-h-auto tw-min-h-[25em] tw-bg-center tw-bg-cover medium:tw-py-24 medium:tw-h-[65vh] large:tw-h-[75vh]" style="background-image:url('{{ hero_background.url }}')">
```
Thanks @mmmavis! 

In my latest commit, I am instead utilizing our preexisting template of `responsive_screen_wide_image.html`, which is also used in `custom_hero.html` and `publication_hero.html` for their hero images, so there is no longer a need for a tailwind class or for updating the inline styles 👍

danielfmiranda commented on network-api/networkapi/wagtailpages/pagemodels/base.py:
```diff
@@ -959,6 +959,7 @@ def get_banner(self):
     ]
 
     subpage_types = [
+        "AppInstallPage",
```
Thanks for double checking @mmmavis!

I am confirming that is the case, based on this request from @kristinashu:

"TikTok url will be [foundation.mozilla.org/TikTokReporter](http://foundation.mozilla.org/TikTokReporter) so it will be a child of the homepage."

danielfmiranda commented:
Thanks for the review everyone!


@kristinashu:
 
> I have a question about the CTA since it's hard to tell what it is in the review app - is the CTA snippet a petition or a newsletter signup? We need to have a newsletter sign up but for future proofing, it might be best if it could be either newsletter signor or petition (like on the existing banner pages).

This new page builds off the existing campaign/banneredcampaign page model, so the CTA section is the same 👍 it can be either a newsletter sign or a petition

> I'm not sure how the top section is being resized but it seems too tall at some browser sizes. It should have a max height and resize like the top section of the Articles pages. For example https://foundation.mozilla.org/en/insights/trustworthy-ai-whitepaper/
Thanks for the feedback! I have updated the template per our conversation in slack, and the hero should now be sized quite similarly to Article pages. 

I have created updated review pages for [tiktok](https://foundation-s-tiktok-pag-1mbxqc.mofostaging.net/en/tiktok/) and [regretsreporter](https://foundation-s-tiktok-pag-1mbxqc.mofostaging.net/en/youtuberegrets/). Can you please take a look whenever you find a free moment and let me know what you think? 


@mmmavis @robdivincenzo, I have updated the PR based on both of your feedback, and it is now ready for re-review. I have tried to leave comments explaining the changes, but if you have any further questions please let me know. Thanks again!

kristinashu reviewed (approved):
So amazing! Thank you @danielfmiranda for getting all the little details of this right!

kristinashu commented:
I'm sorry Daniel, feel free to do this as a follow up ticket. I just got the the final graph from the illustrator and it seems like the image is a little pixelated on the site. Is the image getting compressed by the cms? The original is 383kb which I feel is a reasonable size for a large hero banner. ![AA_FF_1730x567](https://github.com/MozillaFoundation/foundation.mozilla.org/assets/1068783/7e786ef1-4aa3-448b-b907-964d2e42df2e)

robdivincenzo reviewed (approved):
Excellent work @danielfmiranda ! 🔥

simont-cr commented:
@danielfmiranda While reviewing the app, I noticed two issues regarding buttons:

**1. Mobile view buttons width**
When viewing the app on mobile view, the buttons do not have a consistent width. I feel that it would be much better if they are a similar width, maybe matching the bigger button.

![image](https://github.com/MozillaFoundation/foundation.mozilla.org/assets/159044329/ecd8780f-600b-49a2-a5fa-0c8a5b72a911)


2. When using tab for navigation and onFocus, the text turns width on the buttons
Right now when using tab to navigate through the page, whenever the focus is any of the two buttons, the text turns into white but the button stays with the white background.

The onFocus styling should be similar as the onHover, which turns the button background color black.

danielfmiranda commented:
Great catches @simont-cr!

My latest commit to this PR updates the focus state of the buttons 👍 

I am now investigating Kristina's background image issue and the button width issue.

danielfmiranda commented:
Thanks for calling that out @kristinashu! I have created https://mozilla-hub.atlassian.net/browse/TP1-368 to investigate 👍

simont-cr reviewed (approved):
Tabbing through the buttons no longer cause issues and mobile btns are now similar width as expected, looking good, thanks!

mmmavis reviewed (approved):
Thanks @danielfmiranda !

data-sync-user commented:
➤ Simon Acosta Torres commented:

PR has already been merged.
